torch/optim/lars.py

```diff
@@ -0,0 +1,168 @@
+        trust_coefficient: float = 0.001,
+        eps: float = 1e-8,
+        maximize: bool = False,
+        differentiable: bool = False,
+    ):
+        if lr is not required and lr < 0.0:
+            raise ValueError(f"Invalid learning rate: {lr}")
+        if weight_decay < 0.0:
+            raise ValueError(f"Invalid weight decay value: {weight_decay}")
+        if momentum < 0.0:
+            raise ValueError(f"Invalid momentum value: {momentum}")
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+
+        defaults = dict(
+            lr=lr,
+            weight_decay=weight_decay,
+            momentum=momentum,
+            dampening=dampening,
+            nesterov=nesterov,
+            maximize=maximize,
+            differentiable=differentiable,
+        )
+
+        super().__init__(params, defaults)
+        self.trust_coefficient = trust_coefficient
+        self.eps = eps
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("nesterov", False)
+            group.setdefault("maximize", False)
+            group.setdefault("differentiable", False)
+
+    @_use_grad_for_differentiable
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            params_with_grad = []
+            grads = []
+            momentum_buffer_list = []
+
+            for p in group["params"]:
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    grads.append(p.grad)
+
+                    state = self.state[p]
+
+                    momentum_buffer_list.append(state.get("momentum_buffer"))
+
+            lars(
+                params_with_grad,
+                grads,
+                momentum_buffer_list,
+                lr=group["lr"],
+                weight_decay=group["weight_decay"],
+                momentum=group["momentum"],
+                dampening=group["dampening"],
+                nesterov=group["nesterov"],
+                eps=self.eps,
+                trust_coefficient=self.trust_coefficient,
+                maximize=group["maximize"],
+            )
+
+            for p, momentum_buffer in zip(params_with_grad, momentum_buffer_list):
+                state = self.state[p]
+                state["momentum_buffer"] = momentum_buffer
+
+        return loss
+
+
+def lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    weight_decay: float,
+    momentum: float,
+    dampening: float,
+    nesterov: bool,
+    eps: float,
+    trust_coefficient: float,
+    maximize: bool,
+):
+    if not torch.jit.is_scripting():
+        func = _single_tensor_lars
```

Comment:
`func` is not declared if scripting. Is this implementation complete?

Reply:
Only the `single_tensor` implementation is complete. The `multi_tensor` one is still missing.

Now, an exception is raised if scripting.